### PR TITLE
Reflection rates

### PR DESCRIPTION
**Pull request by pomalley**
This is a rebase of `u/rami/reflectionrates` onto master, which was non-trivial due to changes to the deconv server for averaging, dithering, and zeroing. See the discussion at the original PR (#113).

### CODE REVIEW

File: dac_calibration_server.py [hunk 1]
```diff
@@ -264,6 +265,7 @@ def correct_iq(self, c, data, zero_ends=False):
                                                   data,
                                                   loop=c['Loop'],
                                                   zipSRAM=False,
+            calset.setReflection(*c['Reflection'])
```
Comment (maffoo) [inline dac_calibration_server.py]: this looks out of place

File: dac_calibration_server.py [hunk 2]
```diff
@@ -372,6 +374,7 @@ def correct_analog_ft(self, c, data, average_ends=False, dither=False):
 
         calset = yield self.getDACcalset(c)
         calset.setSettling(*c['Settling'])
+            calset.setReflection(*c['Reflection'])
```
Comment (maffoo) [inline dac_calibration_server.py]: indentation is funky

**Comment by maffoo** [comment]
and you're going to add a version 0.5.1 jar as well?

**Comment by pomalley** [comment]
I guess `.jar` is in the gitignore so it I didn't notice that it didn't get added. But there it is.

**Comment by pomalley** [comment]
@rb25 and I fixed up the merge errors there. Can you take another look?

Thanks!

**Comment by maffoo** [comment]
LGTM

**Comment by DanielSank** [comment]
Why is this called reflection "rates". A rate and a reflection seem like different things.

File: ghzdac/correction.py [hunk 3]
```diff
@@ -1162,24 +1185,44 @@ def DACifyFT(self, signal, t0=0, n=8192, offset=0, nfft=None, loop=False,
...
+
+                # Reflections:
+                # add to qubit registry the following keys:
+                # reflectionAmplitudes=[0.05]  #relative amplitude
+                # reflectionRates = [0.01 GHz]    #rate is in GHz, (1/ns)
+                #
+                # Reflections are dealt with by modelling a wire with round-trip time 1/rate, and reflection coefficient amplitude. 
+                # It's the simplest model which can describe the effect of reflections in wiring 
+                # in for example the wiring between the DAC output and fridge ports.
+                # Think about echo, reflections give rise to an endless sum of copies of the original signal with decreasing amplitude:
```
Comment (DanielSank) [inline ghzdac/correction.py]: Long line.

File: ghzdac/correction.py [hunk 4]
```diff
@@ -1162,24 +1185,44 @@ def DACifyFT(self, signal, t0=0, n=8192, offset=0, nfft=None, loop=False,
...
+                # f(t) -> (1-amplitude) Sum_k=0^\infty (amplitude^k f(t-k 1/rate) ).
+                #
+                # Suppose X is an ideal pulse, H the impulse response of a piece of cable (with reflection, settling etc).
```
Comment (DanielSank) [inline ghzdac/correction.py]: Long line.

File: ghzdac/correction.py [hunk 5]
```diff
@@ -1162,24 +1185,44 @@ def DACifyFT(self, signal, t0=0, n=8192, offset=0, nfft=None, loop=False,
...
+                    for rate,amplitude in zip(reflectionRates,reflectionAmplitudes):
+                        if abs(rate) > 0.0:
+                            precalc /= (1.0 - amplitude) / (1.0-amplitude*np.exp(-i_two_pi_freqs/rate))
 
                 
-                #the correction window can have very large amplitudes, therefore the time domain signal can have large oscillations which will be truncated digitally, 
-                #leading to deterioration of the waveform. The large amplitudes in the correction window have low S/N ratios.
-                #Here, we apply a maximum value, i.e. truncate the value, but keep the phase. This way we still have a partial correction, within the limits of the boards. 
-                #Doing it this way also helps a lot with the waveforms being scalable.
+                # The correction window can have very large amplitudes, therefore the time domain signal can have large oscillations which will be truncated digitally,
```
Comment (DanielSank) [inline ghzdac/correction.py]: Long line.

File: ghzdac/correction.py [hunk 6]
```diff
@@ -1162,24 +1185,44 @@ def DACifyFT(self, signal, t0=0, n=8192, offset=0, nfft=None, loop=False,
...
+                            precalc /= (1.0 - amplitude) / (1.0-amplitude*np.exp(-i_two_pi_freqs/rate))
 
                 
-                #the correction window can have very large amplitudes, therefore the time domain signal can have large oscillations which will be truncated digitally, 
-                #leading to deterioration of the waveform. The large amplitudes in the correction window have low S/N ratios.
-                #Here, we apply a maximum value, i.e. truncate the value, but keep the phase. This way we still have a partial correction, within the limits of the boards. 
-                #Doing it this way also helps a lot with the waveforms being scalable.
+                # The correction window can have very large amplitudes, therefore the time domain signal can have large oscillations which will be truncated digitally, 
+                # leading to deterioration of the waveform. The large amplitudes in the correction window have low S/N ratios.
+                # Here, we apply a maximum value, i.e. truncate the value, but keep the phase. This way we still have a partial correction, within the limits of the boards.
```
Comment (DanielSank) [inline ghzdac/correction.py]: Long line.

File: ghzdac/correction.py [hunk 7]
```diff
@@ -986,12 +988,30 @@ def setSettling(self, rates, amplitudes):
...
+        All previously used time constants for the reflections will be replaced.
+        """
+        rates = np.asarray(rates)
+        amplitudes = np.asarray(amplitudes)
+        if np.shape(rates) != np.shape(amplitudes):
+            raise Error('arguments to setReflection must have same shape.')
+        s = np.size(rates)
+        rates = np.reshape(np.asarray(rates),s)
+        amplitudes = np.reshape(np.asarray(amplitudes),s)
+        if (not np.array_equal(self.reflectionRates,rates)) or (not np.array_equal(self.reflectionAmplitudes,amplitudes)):
```
Comment (DanielSank) [inline ghzdac/correction.py]: Long line.

File: dac_calibration_server.py [hunk 8]
```diff
@@ -400,6 +403,14 @@ def setsettling(self, c, rates, amplitudes):
         """
         c['Settling'] = (rates, amplitudes)
 
+    @setting(41, 'Set Reflection', rates=['*v[GHz]: reflection rates'], amplitudes=['*v: reflection amplitudes'])
```
Comment (DanielSank) [inline dac_calibration_server.py]: Long line.

**Comment by DanielSank** [comment]
I marked a few overly long lines. There are more but I don't want to clutter up the comment area too much.

**Comment by DanielSank** [comment]
There are several more. I marked one.

**Pull request closed without merging**